src/openrct2/ride/gentle/Maze.h

```diff
@@ -0,0 +1,180 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2019 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#ifndef _MAZE_H_
+#define _MAZE_H_
+
+#include "../../world/Location.hpp"
+
+#include <array>
+#include <functional>
+#include <optional>
+#include <tuple>
+#include <utility>
+
+/**
+ * Get the new maze subTileIndex as used by Peep::var_37 from the previous subTitleIndex and a direction.
+ * Div the returned value by 4 to get the subTileIndex as used the maze pathfinding methods.
+ */
+extern const uint8_t gMazeGetNewDirectionFromEdge[][4];
+
+/**
+ * Get the bit index of the flag in a mazeEntry indicating the presence of a hedge for a subTileIndex and a direction.
+ */
+extern const uint8_t gMazeCurrentDirectionToHedge[][4];
+
+class MazePathfindingEntry;
+void swap(MazePathfindingEntry& a, MazePathfindingEntry& b);
+
+struct PeepMazeStateRegister
+{
+    uint8_t LastDirection : 2;               // Last direction at the previous subTile
+    uint8_t DirectionAtLastIntersection : 2; // Last direction at the previous intersection
+    uint8_t PeekFlag : 1;                    // True if the guest jumped to check is position
+    uint8_t Unused : 3;
+
+    inline bool checkAndResetPeekFlag()
+    {
+        bool f = PeekFlag;
+        PeekFlag = false;
+        return f;
+    }
+};
+static_assert(sizeof(PeepMazeStateRegister) == sizeof(uint8_t));
+
+/**
+ * Data linked to an intersection perviously encountered by a guest.
+ */
+class MazePathfindingEntry
+{
+    union
+    {
+        struct
+        {
+            uint32_t SubTileIdx : 2;       // Index of the subTile corresponding to the intersection
```

Comment:
Do not use bit masks in the code. It is not multi-platform friendly. You will find that we never use them across the whole codebase for this reason. Instead we have lots of getters and setters. I'm not a fan of stuffing all this information into the struct like this. I would much rather we wait for the new save format and used individual variables of a sensible size.

Reply:
If it's gonna be much longer before the new save file, it might be nice to just find a cross platform way to make it work in the meantime for the sake of allowing it to be merged.

Reply:
Yeah, indeed, the bitfields are implementation dependent, my bad. So I think for the moment reverting aeba1ff and continuing to use bitwise ops, shifts and masks until the new save format. That should fixes the trouble regarding on how the different compilers want to organize the memory and lets only the endianness which is handled by the serialization module.